Sources/WSCore/Extensions/WebSocketExtensionBuilder.swift

```diff
@@ -0,0 +1,151 @@
+    static var name: String { get }
+    /// construct client request header
+    func clientRequestHeader() -> String
+    /// construct server response header based of client request
+    func serverReponseHeader(to: WebSocketExtensionHTTPParameters) -> String?
+    /// construct server version of extension based of client request
+    func serverExtension(from: WebSocketExtensionHTTPParameters) throws -> (any WebSocketExtension)?
+    /// construct client version of extension based of server response
+    func clientExtension(from: WebSocketExtensionHTTPParameters) throws -> (any WebSocketExtension)?
+}
+
+extension WebSocketExtensionBuilder {
+    /// construct server response header based of all client requests
+    public func serverResponseHeader(to requests: [WebSocketExtensionHTTPParameters]) -> String? {
+        for request in requests {
+            guard request.name == Self.name else { continue }
+            if let response = serverReponseHeader(to: request) {
+                return response
+            }
+        }
+        return nil
+    }
+
+    /// construct all server extensions based of all client requests
+    public func serverExtension(from requests: [WebSocketExtensionHTTPParameters]) throws -> (any WebSocketExtension)? {
+        for request in requests {
+            guard request.name == Self.name else { continue }
+            if let ext = try serverExtension(from: request) {
+                return ext
+            }
+        }
+        if let nonNegotiableExtensionBuilder = self as? any _WebSocketNonNegotiableExtensionBuilderProtocol {
+            return nonNegotiableExtensionBuilder.build()
+        }
+        return nil
+    }
+
+    /// construct all client extensions based of all server responses
+    public func clientExtension(from requests: [WebSocketExtensionHTTPParameters]) throws -> (any WebSocketExtension)? {
+        for request in requests {
+            guard request.name == Self.name else { continue }
+            if let ext = try clientExtension(from: request) {
+                return ext
+            }
+        }
+        if let nonNegotiableExtensionBuilder = self as? any _WebSocketNonNegotiableExtensionBuilderProtocol {
+            return nonNegotiableExtensionBuilder.build()
+        }
+        return nil
+    }
+}
+
+/// Protocol for w WebSocket extension that is applied without any negotiation with the other side
+protocol _WebSocketNonNegotiableExtensionBuilderProtocol: WebSocketExtensionBuilder {
+    associatedtype Extension: WebSocketExtension
+    func build() -> Extension
+}
+
+/// A WebSocket extension that is applied without any negotiation with the other side
+public struct WebSocketNonNegotiableExtensionBuilder<Extension: WebSocketExtension>: _WebSocketNonNegotiableExtensionBuilderProtocol {
+    public static var name: String { String(describing: type(of: Extension.self)) }
+
+    let _build: @Sendable () -> Extension
+
+    init(_ build: @escaping @Sendable () -> Extension) {
```

Comment:
Does this need to be `public`? Maybe people want to use non-negotiable extensions mixed with negotiable extensions

Reply:
You can mix non-negotiable extensions with negotiable already eg
```swift
extensions: [
    .nonNegotiatedExtension {
        MyExtension()
    },
    .perMessageDeflate()
]
```